### ISSUE TITLE
channel founder should be able to change the privilege levels of channel operations

### ISSUE THREAD

**Issue by slingamn** [oragono/oragono#1307]
Coming out of discussions with @jlu5. The idea is to let the channel founder configure, via services, what mode is required to set the topic, ban/unban, etc.

Supersedes #1306.

**Comment by slingamn** [comment]
Other stuff that may need to be customizable:

1. Which levels have privileges "over" which levels (e.g., currently ops have privileges over other ops, but halfops don't have it over other halfops and admins don't have it over other admins)
2. Whether having an amode lets you grant amodes to others

**Comment by jlu5** [comment]
Interestingly, if we extend the idea of "let founders set who can change modes" from bans to all channel modes, this will actually be more flexible than most modern IRCds as far as I know. (It'd essentially become a more powerful MLOCK)

**Comment by mogad0n** [comment]
An issue I have is that I like to have very few channel operators "visible" in the channels. Basically I would give `flags +oV mogad0n!*@*` as an example which would basically allow me to have automatic voice and the ability to make myself channel operator when i need which would be using the `/cs op #channel mogad0n` command. Currently that is not possible. 

I remember we discussed about this earlier @slingamn and you mentioned that Channel Founders currently can have any channel mode prefix and still not only can they change their own modes but even without any visible change in their power, they can execute any action that is under the purview of a channel founder which is all good. You mentioned that if a channel operator is not visible and then they enact a ban that would make it difficult for a channel founder to diagnose or for the other users to hold accountable, but in the flags system you cannot do things a channel operator can without making yourself operator doing the deed and then going back to pleb status.

The following suggestion would solve what I would like but unless i misunderstood it, it has flaws:

> "Interestingly, if we extend the idea of "let founders set who can change modes" from bans to all channel modes, this will actually be more flexible than most modern IRCds as far as I know. (It'd essentially become a more powerful MLOCK)"  by @jlu5

One of issues this has is the same accountability; and tracking the changes and who made them problem. If you start giving individuals (accounts) specific "capabilities" then it can become a mess to keep track of stuff you've handed out after some time has gone by. Secondly, the entire point of these channel mode prefixes becomes redundant as seeing them in a channel has no meaning. What you could do is have the ability to configure what having an +a or an +o means. 

So to conclude it would be nice if "persistent modes" include the ability to stay on lower levels and that respects part/joins but for actions that require higher cmodeprefs you have to `/op` yourself as an example in order to execute them.

**Comment by DanielOaks** [comment]
Honestly, this is one of the features that I see implemented in anope and atheme and I don't really find much use for? At least in my experience it never has been, really. Configurability is cool, but it does make things easier if there's a defined set of who can do what that's always true, that we can list in the manual or help files, and people can sort their channel moderation staff according to those priv levels.

The _"Which levels have privileges "over" which levels"_ thing is something that IRC has generally always standardised on - the alternative to this is honestly implement Anope's levels system where each chanop gets a level number assigned to them and that can be used to decide who has power over who. I think the default settings for these (halfop can't kick/etc halfop, +a can't kick/etc +a) both make a fair deal of sense - halfops will go to chanops for disputes, +a will go to founders for disputes.

On the _"Whether having an amode lets you grant amodes to others"_ topic, I think that you typically can do so? I'm pretty fine with this, generally. Voiced aren't able to give other people automatatic-voice, but that's because they're not really channel staff in terms of moderation privs.

The real use for this is for specific channel founders that want to fine-tune the moderation levels on their channel. For that, I think redefining the modes themselves (`+qaohv`) gets pretty confusing and means we can't document it as well imo, and if we really want to go that way it might be best to just look at levels instead. But I'm not a big fan of customising this because I've never found it very useful myself.

**Comment by mogad0n** [comment]
I kind of agree that at this stage you shouldn't redefine the modes themselves and also adding individual capabilities has problems.

I only mentioned my issue here because i wanted to 1) not add another issue when there is an open one which is somewhat related 2) It may possibly help arrive at a conclusion for this one as well

What I wanted to sort of see is:
1) Setting a persistent mode != automoding said users when they join a channel.
2) You can have the persistent mode `+O` which auto grants chanops whenever you join and  `+o` which means you have the same persistent mode really it's just that you aren't autochanopped whenever you join.
3) But this will possibly need you to be able to grant non-singular amount of persistent modes to an account. As you'd usually have `+oV` which means grant voice upon join and they can `/cs op` themselves to enact say bans or whatnot.

@DanielOaks just to understand your response better what problems do you see in specifically these points above. This is possible in the major services but not here. Here if you have `+o` you will get auto op'd upon.

**Comment by DanielOaks** [comment]
I don't see any issues with the above points, but this issue is specifically about the privilege levels of actions. I like the flags system on atheme/freenode, works pretty nicely once you contend with the fact that you need to go messaging chanserv (I hate any feature that requires messaging nickserv/chanserv/etc, so I prefer the current set-the-modes-and-forget way we have things now as the default method of giving users `qaohv`).

**Comment by slingamn** [comment]
>An issue I have is that I like to have very few channel operators "visible" in the channels.

1. Go always-on
2. Have an AMODE
3. Remove the actual mode with MODE; this removal is sticky because you're always-on
4. If you need ops, you can get them back with JOIN/PART

Here's a thought, though: `CS OP` could regrant your AMODE as a MODE (right now only the founder can use it, and it unconditionally regrants +q to the founder and +o to any other user). (If this works for your use case, I'll implement this but not in 2.5, since we're already in the release window; you're running master anyway.)

**Comment by mogad0n** [comment]
Yes, there is no hurry to do this. Maybe `CS OP` could be `CS MODE`. I can see issues and inconsistent workflows as the issue here . We can revisit this later, as the Services develop.
